### PR TITLE
WIP: workload: support pgbouncer transaction pooling with tpcc

### PR DESCRIPTION
**Pull request by RichardJCai**
Allow initialization of tpcc with no prepared statements.

Release note: None

### CODE REVIEW

File: pkg/workload/cli/run.go [hunk 1]
```diff
@@ -52,6 +52,8 @@ var ramp = runFlags.Duration("ramp", 0*time.Second, "The duration over which to
 
 var initFlags = pflag.NewFlagSet(`init`, pflag.ContinueOnError)
 var drop = initFlags.Bool("drop", false, "Drop the existing database, if it exists")
+var runInsertInTx = initFlags.Bool("run-insert-in-tx", false,
```
Comment (rafiss) [inline pkg/workload/cli/run.go:55]: why did we need the inserts in a txn for this?

File: pkg/workload/pgx_helpers.go [hunk 2]
```diff
@@ -39,6 +39,9 @@ type MultiConnPoolCfg struct {
 	// If 0, there is no per-pool maximum (other than the total maximum number of
 	// connections which still applies).
 	MaxConnsPerPool int
+
+	// NoPrepare determines whether or not prepared statements can be used or not.
+	NoPrepare bool
```
Comment (rafiss) [inline pkg/workload/pgx_helpers.go:44]: sometimes i find boolean flags that have a negation in the name more confusing. could this just be named `UseSimpleProtocol` or something?

File: pkg/workload/sql_runner.go [hunk 3]
```diff
@@ -47,22 +47,36 @@ type SQLRunner struct {
...
-	noprepare
-	simple
+	// Prepare: we prepare the query once during Init, then we reuse it for
+	// each execution. This results in a Bind and Execute on the server each time
+	// we run a query (on the given connection). Note that it's important to
+	// prepare on separate connections if there are many parallel workers; this
+	// avoids lock contention in the sql.Rows objects they produce. See #30811.
+	Prepare Method = iota
+	// NoPrepare: each query is issued separately (on the given connection).
+	// This results in Parse, Bind, Execute on the server each time we run a
```
Comment (rafiss) [inline pkg/workload/sql_runner.go:66]: i'm pretty sure the driver will automatically prepare it and reuse the stmt, even if we don't use it by name. anyway, no need to change this now, that's beyond the scope of this PR.

although hmm maybe we should just get rid of noprepare entirely. it doesn't seem helpful really.

File: pkg/workload/tpcc/tpcc.go [hunk 4]
```diff
@@ -620,13 +621,22 @@ func (w *tpcc) Ops(
 	w.reg = reg
 	w.usePostgres = parsedURL.Port() == "5432"
 
+	method, err := workload.StringToMethod(w.connFlags.Method)
+	if err != nil {
+		return workload.QueryLoad{}, err
+	}
+
+	// We only Prepare statements if Prepare is explicitly specified.
+	noPrepare := method != workload.Prepare
```
Comment (rafiss) [inline pkg/workload/tpcc/tpcc.go:630]: i think the condition should be `method == workload.Simple` (based on my previous comment)

**Review by rafiss — CHANGES_REQUESTED** [review]
i think this is almost ready! let's try to get this in

**Comment by dbist** [comment]
would it be possible to backport to 21.1?

**Pull request closed without merging**